modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryFutureAdapter.java

```diff
@@ -324,8 +242,8 @@ public void onPage(@Nullable UUID nodeId, @Nullable Collection<?> data, @Nullabl
 
         try {
             if (err != null)
-                synchronized (this) {
-                    enqueue(Collections.emptyList());
+                synchronized (lock) {
+                    reducer().onError();
```

Comment:
We still need synchronization there, as it's required to change reducer queue atomically with future onDone. If user hands on latch await, we has to count down it after future is done. Otherwise we can miss exception, we have tests for this logic. Please check this comment:
https://github.com/apache/ignite/blob/6b6c32bf1582d50f349f522c04b11d1603049797/modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryFutureAdapter.java#L330

We can control future.onDone() within reducer, but I think it isn't responsibility of reducer.